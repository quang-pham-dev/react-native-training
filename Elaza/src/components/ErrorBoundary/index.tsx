import React, {Component, PropsWithChildren, ErrorInfo} from 'react'
import {ScrollView, Text} from 'react-native'

export enum ErrorMode {
  ALWAYS,
  DEV,
  PROD,
  NEVER,
}

interface Props {
  errorMode: ErrorMode
}

interface State {
  error?: Error
  errorInfo?: ErrorInfo
}

class ErrorBoundary extends Component<PropsWithChildren<Props>, State> {
  state: State = {}

  // If an error in a child is encountered, this will run
  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    // Catch errors in any components below and re-render with error message
    this.setState({
      error,
      errorInfo,
    })

    // You can also log error messages to an error reporting service here
    // This is a great place to put BugSnag, Sentry, Honeybadger, etc:
    // reportErrorToCrashReportingService(error)
  }

  // Reset the error back to null
  resetError = () => {
    this.setState({})
  }

  // To avoid unnecessary re-renders
  shouldComponentUpdate(
    nextProps: Readonly<any>,
    nextState: Readonly<any>,
  ): boolean {
    return nextState.error !== nextProps.error
  }

  // Only enable if we're catching errors in the right environment
  isEnabled(): boolean {
    return (
      this.props.errorMode === ErrorMode.ALWAYS ||
      (this.props.errorMode === ErrorMode.DEV && __DEV__) ||
      (this.props.errorMode === ErrorMode.PROD && !__DEV__)
    )
  }

  // Render an error UI if there's an error; otherwise, render children
  render() {
    return this.isEnabled() && this.state.error ? (
      <ScrollView>
        <Text>{`${this.state.error}`}</Text>
        {this.state.errorInfo && (
          <Text>{`${this.state.errorInfo.componentStack}`}</Text>
        )}
      </ScrollView>
    ) : (
      this.props.children
    )
  }
}

export default ErrorBoundary
