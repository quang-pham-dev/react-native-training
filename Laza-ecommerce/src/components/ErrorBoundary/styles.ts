import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },

  subContainer: {
    marginHorizontal: 30,
    alignItems: 'center'
  },

  icon: {
    marginBottom: 18
  },

  bigBoldText: {
    fontSize: 18,
    color: 'black',
    letterSpacing: 0.8,
    fontWeight: 'bold'
  }
});

export default styles;
