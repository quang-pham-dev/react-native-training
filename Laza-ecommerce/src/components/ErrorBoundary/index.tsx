import React, { Component } from 'react';
import { Text, View } from 'react-native';

// Components
import Title from 'components/Title';

// Styles
import styles from './styles';
import { BOUNDARY_ERROR_MESSAGE } from 'constants/Common';

export default class ErrorBoundary extends Component {
  state = {
    hasError: false,
    error: null
  };

  static getDerivedStateFromError(error: any) {
    return { hasError: true, error };
  }

  render() {
    if (this.state.hasError) {
      return (
        <View style={styles.container}>
          <View style={styles.subContainer}>
            <Title.HeadingPage
              style={styles.bigBoldText}
              titleName={BOUNDARY_ERROR_MESSAGE}
              titleAlign='center'
            />
          </View>
        </View>
      );
    }

    return this.props.children;
  }
}
