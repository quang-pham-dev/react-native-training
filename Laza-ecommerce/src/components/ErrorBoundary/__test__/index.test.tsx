import React from 'react';

// LIBS
import renderer from 'react-test-renderer';

// Components
import ErrorBoundary from 'components/ErrorBoundary';
// Themes

describe('ErrorBoundary Component', () => {
  const ChildrenComponent = () => {
    throw new Error('Something when wrong');
  };

  const tree = renderer.create(
    <ErrorBoundary>
      <ChildrenComponent />
    </ErrorBoundary>
  );

  test('should render correctly', () => {
    const component = tree.toJSON();
    expect(component).toMatchSnapshot();
  });
});
