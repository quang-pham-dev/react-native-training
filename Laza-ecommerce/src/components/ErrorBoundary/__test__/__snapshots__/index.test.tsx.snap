// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Component should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginHorizontal": 30,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "Inter_500Medium",
            "fontSize": 17,
            "lineHeight": 16,
            "textTransform": "none",
          },
          Array [
            Object {
              "fontFamily": "Inter_600SemiBold",
              "fontSize": 28,
              "lineHeight": 30,
            },
            Object {
              "color": "black",
              "fontSize": 18,
              "fontWeight": "bold",
              "letterSpacing": 0.8,
            },
          ],
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      Something went wrong!
    </Text>
  </View>
</View>
`;
